tests/ui/cast.rs

```diff
@@ -61,4 +61,15 @@ fn main() {
     1u32 as usize; // Should not trigger any lint
     1i32 as isize; // Neither should this
     1i32 as usize;
+
+    // Test cast_unnecessary
+    1i32 as i32;
+    1f32 as f32;
+    false as bool;
+    &1i32 as &i32;
+
+    1i32 as i64;      // Should not trigger
```

Comment:
`expr_ty_adjusted` and `expr_ty` seems to be the same for `1 as u64`:
```
let (cast_from, cast_to) = (cx.tables.expr_ty(ex), cx.tables.expr_ty(expr));
            let (cast_from_adj, cast_to_adj) = (cx.tables.expr_ty_adjusted(ex), cx.tables.expr_ty_adjusted(expr));
            println!("1. {:?} {:?} {:?} {:?}", cast_from_adj, cast_from_adj.sty, cast_to_adj, cast_to_adj.sty);
            println!("2. {:?} {:?} {:?} {:?}", cast_from, cast_from.sty, cast_to, cast_to.sty);
```
prints:
```
1. u64 TyUint(u64) u64 TyUint(u64)
2. u64 TyUint(u64) u64 TyUint(u64)
```